File: backend/internal/handler/immich.go
package handler

import (
	"net/http"
	"strconv"

	"github.com/aitjcize/esp32-photoframe-server/backend/internal/service"
	"github.com/labstack/echo/v4"
)

type ImmichHandler struct {
	immich *service.ImmichService
}

func NewImmichHandler(s *service.ImmichService) *ImmichHandler {
	return &ImmichHandler{immich: s}
}

// serverRequest is the add/edit body for an Immich server. An empty APIKey on
// edit means "keep the existing key" (the UI never echoes secrets back).
type serverRequest struct {
	Label   string `json:"label"`
	URL     string `json:"url"`
	APIKey  string `json:"api_key"`
	Enabled bool   `json:"enabled"`
}

// TestConnection tests arbitrary credentials (the add/edit form's "Test" button,
// before the server is saved). Body: {url, api_key}.
func (h *ImmichHandler) TestConnection(c echo.Context) error {
	var req serverRequest
	if err := c.Bind(&req); err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid request"})
	}
	if err := h.immich.TestCredentials(req.URL, req.APIKey); err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": err.Error()})
	}
	return c.JSON(http.StatusOK, map[string]string{"status": "ok"})
}

// ListServers returns all configured Immich servers.
func (h *ImmichHandler) ListServers(c echo.Context) error {
	servers, err := h.immich.AllServers()
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	return c.JSON(http.StatusOK, servers)
}

func (h *ImmichHandler) CreateServer(c echo.Context) error {
	var req serverRequest
	if err := c.Bind(&req); err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid request"})
	}
	srv, err := h.immich.CreateServer(req.Label, req.URL, req.APIKey, req.Enabled)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	return c.JSON(http.StatusOK, srv)
}

func (h *ImmichHandler) UpdateServer(c echo.Context) error {
	id, _ := strconv.Atoi(c.Param("id"))
	var req serverRequest
	if err := c.Bind(&req); err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid request"})
	}
	if err := h.immich.UpdateServer(uint(id), req.Label, req.URL, req.APIKey, req.Enabled); err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	return c.JSON(http.StatusOK, map[string]string{"status": "updated"})
}

func (h *ImmichHandler) DeleteServer(c echo.Context) error {
	id, _ := strconv.Atoi(c.Param("id"))
	if err := h.immich.DeleteServer(uint(id)); err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": err.Error()})
	}
	return c.JSON(http.StatusOK, map[string]string{"status": "deleted"})
}

// TestServer tests a saved server by id.
func (h *ImmichHandler) TestServer(c echo.Context) error {
	id, _ := strconv.Atoi(c.Param("id"))
	if err := h.immich.TestConnection(uint(id)); err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": err.Error()})
	}
	return c.JSON(http.StatusOK, map[string]string{"status": "ok"})
}

// ListAlbums returns the albums exposed by every enabled server, tagged with the
// server id + label so the per-device picker can group by server.
func (h *ImmichHandler) ListAlbums(c echo.Context) error {
	albums, err := h.immich.ListAllAlbums()
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	return c.JSON(http.StatusOK, albums)
}

// Sync runs the non-destructive incremental sync (new/removed/edited assets,
// stable IDs, rotation cursors preserved) across all servers. "Sync Now" button.
func (h *ImmichHandler) Sync(c echo.Context) error {
	if err := h.immich.SyncNow(); err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	return c.JSON(http.StatusOK, map[string]string{"status": "synced"})
}

// Resync hard-deletes all Immich photos and re-imports from scratch. Destructive:
// resets every frame's rotation cursor and re-fetches people/faces. Backs the
// "Rebuild Library" button (UI confirms first).
func (h *ImmichHandler) Resync(c echo.Context) error {
	if err := h.immich.ClearAndResync(); err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	return c.JSON(http.StatusOK, map[string]string{"status": "resynced"})
}

func (h *ImmichHandler) Clear(c echo.Context) error {
	if err := h.immich.ClearPhotos(); err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	return c.JSON(http.StatusOK, map[string]string{"status": "cleared"})
}

func (h *ImmichHandler) GetPhotoCount(c echo.Context) error {
	count, err := h.immich.GetPhotoCount()
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	return c.JSON(http.StatusOK, map[string]interface{}{"count": count})
}

// UsedAlbums returns the Immich albums that currently have synced photos, for
// the Gallery's per-album filter (see ListUsedAlbums).
func (h *ImmichHandler) UsedAlbums(c echo.Context) error {
	albums, err := h.immich.ListUsedAlbums()
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	return c.JSON(http.StatusOK, albums)
}

File: webapp/src/stores/immich.ts
import { defineStore } from 'pinia';
import { api } from '../api';

export interface ImmichServer {
  id: number;
  label: string;
  url: string;
  api_key?: string;
  enabled: boolean;
}

export const useImmichStore = defineStore('immich', {
  state: () => ({
    count: 0,
    // albums are AlbumInfo: { id, name, server_id, server_label }
    albums: [] as any[],
    servers: [] as ImmichServer[],
    loading: false,
    error: null as string | null,
  }),
  actions: {
    async fetchServers() {
      try {
        const res = await api.get('/immich/servers');
        this.servers = res.data || [];
      } catch (e: any) {
        console.error('Failed to fetch Immich servers', e);
      }
    },

    async createServer(payload: Partial<ImmichServer>) {
      await api.post('/immich/servers', payload);
      await this.fetchServers();
    },

    async updateServer(id: number, payload: Partial<ImmichServer>) {
      await api.put(`/immich/servers/${id}`, payload);
      await this.fetchServers();
    },

    async deleteServer(id: number) {
      await api.delete(`/immich/servers/${id}`);
      await this.fetchServers();
    },

    // Test a saved server by id.
    async testServer(id: number) {
      const res = await api.post(`/immich/servers/${id}/test`);
      return res.data;
    },

    // Test arbitrary credentials before saving (add/edit form).
    async testCredentials(url: string, apiKey: string) {
      const res = await api.post('/immich/test', { url, api_key: apiKey });
      return res.data;
    },

    async fetchCount() {
      try {
        const res = await api.get('/immich/count');
        this.count = res.data.count || 0;
      } catch (e: any) {
        console.error('Failed to fetch Immich photo count', e);
      }
    },

    async fetchAlbums() {
      this.loading = true;
      this.error = null;
      try {
        const res = await api.get('/immich/albums');
        this.albums = res.data;
      } catch (e: any) {
        this.error = e.response?.data?.error || e.message;
        throw e;
      } finally {
        this.loading = false;
      }
    },

    // Test the default server's credentials. Pass the entered url + key so the
    // check hits exactly what's on the form (the /immich/test endpoint tests the
    // credentials in the body).
    async testConnection(url?: string, apiKey?: string) {
      this.loading = true;
      try {
        const body = url && apiKey ? { url, api_key: apiKey } : {};
        const res = await api.post('/immich/test', body);
        await this.fetchCount();
        await this.fetchServers();
        await this.fetchAlbums().catch(() => {});
        return res.data;
      } catch (e: any) {
        throw e;
      } finally {
        this.loading = false;
      }
    },

    async sync() {
      this.loading = true;
      try {
        await api.post('/immich/sync');
        await this.fetchCount();
      } catch (e: any) {
        throw e;
      } finally {
        this.loading = false;
      }
    },

    async resync() {
      this.loading = true;
      try {
        await api.post('/immich/resync');
        await this.fetchCount();
      } catch (e: any) {
        throw e;
      } finally {
        this.loading = false;
      }
    },
  },
});
